From d3e4b8cf2ba322efe1db2acbe411dfc27d46ac7c Mon Sep 17 00:00:00 2001
From: Jerome Brunet <jbrunet@baylibre.com>
Date: Mon, 22 Nov 2021 11:54:04 +0100
Subject: [PATCH 1/2] mmc: meson-gx: stick to the oscillator

This is a temporary workaround to make things stable during boot time

There seems to be problem on chips from the g12a generation.
Some work with a core phase of 180, other of 270.
It is difficult to get a common setting which makes sense.

It worth mentioning that Linux handle the all devices with a 180
degree core phase without issue so far. So does the ROMCode.

There must be something odd with u-boot.
At 25MHz on FDIV2 it fails to select a mode on some devices
At 24MHz on the OSC, it succeed

Until this is fixed let's stick to the OSC

Signed-off-by: Jerome Brunet <jbrunet@baylibre.com>
---
 drivers/mmc/meson_gx_mmc.c | 37 +++++++++++++++----------------------
 drivers/mmc/meson_gx_mmc.h |  4 ++--
 2 files changed, 17 insertions(+), 24 deletions(-)

diff --git a/drivers/mmc/meson_gx_mmc.c b/drivers/mmc/meson_gx_mmc.c
index a5e9ac5637b9..fa0bd379db72 100644
--- a/drivers/mmc/meson_gx_mmc.c
+++ b/drivers/mmc/meson_gx_mmc.c
@@ -52,28 +52,21 @@ static void meson_mmc_config_clock(struct mmc *mmc)
 
 	/* TOFIX This should use the proper clock taken from DT */
 
-	/* 1GHz / CLK_MAX_DIV = 15,9 MHz */
-	if (mmc->clock > 16000000) {
-		clk = SD_EMMC_CLKSRC_DIV2;
-		clk_src = CLK_SRC_DIV2;
-	} else {
-		clk = SD_EMMC_CLKSRC_24M;
-		clk_src = CLK_SRC_24M;
-	}
-	clk_div = DIV_ROUND_UP(clk, mmc->clock);
-
 	/*
-	 * SM1 SoCs doesn't work fine over 50MHz with CLK_CO_PHASE_180
-	 * If CLK_CO_PHASE_270 is used, it's more stable than other.
-	 * Other SoCs use CLK_CO_PHASE_180 by default.
-	 * It needs to find what is a proper value about each SoCs.
+	 * FIXME:
+	 * keep clock on the oscillator. Some board fail initialize MMC with FDIV.
+	 * U-boot fails to select a mode with this clock
+	 * Some work well with core phase at 180, other at 270. This is odd and indicate
+	 * another problem we need to deal with
 	 */
-	if (meson_gx_mmc_is_compatible(mmc->dev, MMC_COMPATIBLE_SM1))
-		meson_mmc_clk |= CLK_CO_PHASE_270;
-	else
-		meson_mmc_clk |= CLK_CO_PHASE_180;
+	clk = SD_EMMC_CLKSRC_24M;
+	clk_src = CLK_SRC_24M;
+	clk_div = DIV_ROUND_UP(clk, mmc->clock);
+
+	/* 180 phase core clock */
+	meson_mmc_clk |= CLK_CO_PHASE_180;
 
-	/* 180 phase tx clock */
+	/* 0 phase tx clock */
 	meson_mmc_clk |= CLK_TX_PHASE_000;
 
 	/* clock settings */
@@ -326,9 +319,9 @@ int meson_mmc_bind(struct udevice *dev)
 }
 
 static const struct udevice_id meson_mmc_match[] = {
-	{ .compatible = "amlogic,meson-gx-mmc", .data = MMC_COMPATIBLE_GX },
-	{ .compatible = "amlogic,meson-axg-mmc", .data = MMC_COMPATIBLE_GX },
-	{ .compatible = "amlogic,meson-sm1-mmc", .data = MMC_COMPATIBLE_SM1 },
+	{ .compatible = "amlogic,meson-gx-mmc", .data = MMC_COMPATIBLE_V2 },
+	{ .compatible = "amlogic,meson-axg-mmc", .data = MMC_COMPATIBLE_V3 },
+	{ .compatible = "amlogic,meson-sm1-mmc", .data = MMC_COMPATIBLE_V3 },
 	{ /* sentinel */ }
 };
 
diff --git a/drivers/mmc/meson_gx_mmc.h b/drivers/mmc/meson_gx_mmc.h
index 92aec5329f6e..ac1906d2ea47 100644
--- a/drivers/mmc/meson_gx_mmc.h
+++ b/drivers/mmc/meson_gx_mmc.h
@@ -10,8 +10,8 @@
 #include <linux/bitops.h>
 
 enum meson_gx_mmc_compatible {
-	MMC_COMPATIBLE_GX,
-	MMC_COMPATIBLE_SM1,
+	MMC_COMPATIBLE_V2,
+	MMC_COMPATIBLE_V3,
 };
 
 #define SDIO_PORT_A			0
-- 
2.34.0

